dashboard/app/helpers/openai_evaluate_helper.rb

```diff
@@ -65,6 +67,95 @@ def self.evaluate(level, unit, options)
     end
   end
 
+  def self.evaluate_section(unit, section)
+    students = section.students
+
+    levels = unit.levels
+
+    user_levels = UserLevel.joins(:user, :level).
+                          where(user: students, level: levels, script: unit).
+                          includes(:user, :level, :level_source)
+
+    user_levels.each do |user_level|
+      if user_level.level_source && user_level.level_source.data.present?
+        evaluate_free_response(user_level, unit)
+      elsif user_level.level.name == 'U4 L03 Variables operator practice 5_2024' || user_level.level.name == 'U4 L03 Variables numbers practice 4_2024'
+        evaluate_code_level(user_level, unit)
+      end
+    end
+
+    nil
+  end
+
+  def self.evaluate_free_response(user_level, unit)
+    student_work = user_level.level_source.data
+
+    response = evaluate(
+      user_level.level,
+      unit,
+      student_work: student_work,
+      evaluation_type: SharedConstants::AI_EVALUATION_TYPES[:SINGLE_STUDENT]
+    )
+
+    create_ai_evaluations_from_ai_response(user_level.user, user_level, unit, response, {})
+  end
+
+  def self.evaluate_code_level(user_level, unit)
+    # Why isn't this retrieving the student code?
+    helper = ApplicationController.helpers
+    student_code = helper.get_student_code(user_level.user.id, user_level.level, unit.id)
```

Comment:
Can just call it as `get_student_code`. I don't think you  need the `ApplicationController.helpers` bit... but the only other places we call this are in controllers not helpers so I could be wrong.

Reply:
This wasn't working, I think because we are not in a controller. This is the only bit of code I could find that worked, but happy to keep looking if you prefer.

Reply:
If it works, go for it! I'm not worried about it.